keras_nlp/models/xlnet/xlnet_backbone.py

```diff
@@ -0,0 +1,262 @@
+            `[batch_size, sequence_length]`.
+        padding_mask: Mask to avoid performing attention on padding token indices of shape `[batch_size, sequence_length]`.
+        target_mapping: Optional `Tensor` which denotes mask to indicate the output tokens to use of shape
+            `[batch_size, num_predict, sequence_length]`. If `target_mapping[k, i, j] = 1`, the i-th predict in batch k
+            is on the j-th token. Only used during pretraining.
+        perm_mask: Optional `Tensor` which denotes mask to indicate the attention pattern for each input token of shape
+            `[batch_size, sequence_length, sequence_length]`. Only used during pretraining.
+        mems: Optional `Tensor` of shape `[batch_size, sequence_length(of previous hidden state), hidden_dim, num_layers]`
+            to denote the previous hidden states. If passed, this is also attended over as in Transformer-XL.
+
+    Returns:
+        last_hidden_state: last hidden state of query state of shape `[batch_size, num_predict, hidden_dim]` if query state is not None
+            otherwise last hidden state of content of shape `[batch_size, sequence_length, hidden_dim]`.
+        new_mems: new memory units returned by the model. These are the conatenated
+            tensors of previous mems and hidden states of most recent pass.
+
+    Examples:
+    ```python
+    input_data = {
+        "token_ids": tf.constant(
+            [460, 5272, 1758, 4905, 9, 4, 3], shape=(1, 7),
+        ),
+        "token_type_ids": tf.constant(
+            [0, 0, 0, 0, 0, 0, 2], shape=(1, 7),
+        ),
+        "padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1], shape=(1, 7)
+        ),
+        "target_mapping": tf.random.uniform(minval=0, maxval=2, shape=(1, 5, 7), dtype=tf.int64),
+        "perm_mask": tf.random.uniform(minval=0, maxval=2, shape=(1, 7, 7), dtype=tf.int64),
+        "mems": tf.random.uniform((12, 7, 1, 768), dtype=tf.float64)
+    }
+
+    # Randomly initialized XLNet encoder with a custom config
+    model = keras_nlp.models.XLNetBackbone(
+        vocabulary_size=32000,
+        num_layers=12,
+        num_heads=12,
+        hidden_dim=768,
+        intermediate_dim=3072,
+    )
+    output = model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.0,
+        activation="gelu",
+        kernel_initializer_range=0.02,
+        bias_initializer="zeros",
+        layer_norm_epsilon=1e-12,
+        **kwargs,
+    ):
+        # Inputs
+        token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="token_ids"
+        )
+        padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+        token_type_id = keras.Input(
+            shape=(None,), dtype="int32", name="token_type_ids"
+        )
+        mems = tf.keras.Input(
```

Comment:
`mems` looks like a key value cache, is that correct? what else is in there?

Potentially we might want to rewrite this so that the only backbone inputs are `token_ids`, `padding_mask` and `segment_ids` (like bert). The issue with these inputs is they are going to be needed any time you call the backbone in any context.

If you write a language modeling class, you could write a separate forward pass that take in the `cache`, see https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/gpt2/gpt2_causal_lm.py. Probably also better to call this `cache` instead of `mems` if I'm understanding it's purpose correctly.

For below, we could add a separate pretraining task that take in `perm_mask` and `target_mapping`. This could reasonably be something we leave off for the initial release, as a lot more people will be using the pre-trained checkpoints than attempting to pre-train it.

Reply:
Hi @mattdangerw thanks for you comments! and for raising this concern about the optional inputs, actually `mems` is not key value cache that we use in gpt, its actually inspired from `Transformer-XL`. 

Just to explain in brief - Suppose that we want to process very very large files which do not fit in our context for single forward pass, then we divide the text into smaller segments and process it in different forward passes, but there is one problem... Suppose that we are doing a Question Answering task, and the answer we want to find lies in the previous segment, but since the model does not have access to the previous forward pass data for the present forward pass, the model fails in the task! 
In order to add the information from the previous segments, we pass the `hidden_states` from the previous pass(actually we concatenate the previous hidden_state with the new ones) and call the previous hidden_states as `mems`. 

I think its better to call this `mems` or `states` and not to confuse with `key_value_cache`.
This `mems` is optional but people may want to use `Transformer-xl` or `xl-net` because they have this feature. 

So just to wrap up I will change the design that only allows `token_ids`, `padding_mask` and `segment_ids` to pass through and create an additional forward pass that take in the `mems`.

Reply:
Thanks this is very helpful. Do you have an idea of how to use `mems` in a code sample? Pseudo code is fine.

I'm trying to wrap my head around how this would be supported via our task API (or maybe we don't). Potentially one thing we could do is expose this as a separate forward pass. Much like we have `call_with_cache` for decoder models. So if you are doing something simple where everything fits in the context window (which is probably the bulk of users), this mems stuff does not get in your way. But if you are doing something with long documents, you have a workflow available to you.

Reply:
I think it should look like this - 


- First we tokenize both of the segments
```
input_tokens_1 = tokenizer.tokenize(text_segment_1)
input_tokens_2 = tokenizer.tokenize(text_segment_2)
```

- We define the classifier(To be implemented after this PR is merged)
```
xlnet_classifier = XLNetClassifier(...)
```

- Get mems_output from the first segment's forward pass
```
last_hidden_state1, mems_output1 = xlnet_classifier(input_tokens_1)
```

- Use the previous mems(from input_tokens_1) as prev_mems in the new `call_with_mems` 
```
last_hidden_state2, mems_output2 = xlnet_classifier.call_with_mems(input_tokens_2, 
                                                                   prev_mems=mems_output1)
```

Please note that I will work on Classifier and Tokenizer after this PR is merged. Also currently XLNetBackbone does not output `mems`, but will be updated in the future commit.